<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <!--<meta http-equiv="X-UA-Compatible" content="IE=edge">-->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="shortcut icon" href="/img/favicon.png">
    <link rel="alternate" type="application/rss+xml" title="RSS" href="http://www.nextflow.io/feed.xml">
    <title>Nextflow</title>
    <meta name="keywords" content="dataflow, data-flow, dataflow programming model, flow-based programming, pipeline, computational pipeline, bioinformatics, bioinformatics pipeline, RNAseq, next gen sequencing, next generation sequencing, parallel computing, cloud computing, distributed computing, grid computing, cluster computing, Sun Grid Engine, SGE, Oracle Grid Engine"/>

    <!-- Bootstrap core CSS -->
    <link href="/css/bootstrap.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="/css/color-styles.css" rel="stylesheet">
    <link href="/css/ui-elements.css" rel="stylesheet">
    <link href="/css/custom.css" rel="stylesheet">

    <!-- Resources -->
    <link href="/css/animate.css" rel="stylesheet">
    <link href="http://netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css" rel="stylesheet">
    <link href='http://fonts.googleapis.com/css?family=Lobster' rel='stylesheet' type='text/css'>

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
    <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

</head>

<body class="body-green">

<div class="navbar navbar-inverse navbar-fixed-top" role="navigation">

    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand" href="/index.html" style='padding: 8px'><img src='/img/nextflow2014_no-bg.png' height='40px' style='border: 0;'></a>
        </div>
        <div class="navbar-collapse collapse">

            <ul class="nav navbar-nav navbar-left hidden-xs">
                <li class="show animated flipInX" >
                    <a href="/index.html#Features" class="pull-right">Features</a>
                </li>
                <li class="show animated flipInX" >
                    <a href="/index.html#GetStarted" class="pull-right">Quick start</a>
                </li>

                <li class="dropdown show animated flipInX">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Examples <b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        <li><a href="/example1.html">Basic pipeline</a></li>
                        <li><a href="/example2.html">Mixing scripting languages</a></li>
                        <li><a href="/example3.html">BLAST pipeline</a></li>
                        <li><a href="/example4.html">RNAseq pipeline</a></li>
                    </ul>
                </li>

                <li class="show animated flipInX">
                    <a href="/docs/latest/index.html" class="pull-right" target="_blank">Documentation</a>
                </li>

                <li class="show animated flipInX">
                    <a href="/blog.html">Blog</a>
                </li>

                <li class="show animated flipInX">
                    <a href="/about-us.html">About Us</a>
                </li>
            </ul>
        </div>
    </div>
</div>
	
	
	<!-- Body -->
	<div class="wrapper"> <!-- wrapper -->
	  <div class="container">
	    <div class="row">
		  <div class="col-xs-12">
		    <h3>Blog <small>Posts or news stories regarding Nextflow, computational pipelines and parallel computing</small></h3>
		    <hr>
		  </div>
		</div>
	    <div class="row">
		  <div class="col-sm-8">
			<div class="timeline">
			  <div class="blg-summary">
			    <h3 ><a href="blog/2014/using-docker-in-hpc-cluster.html">Using Docker for scientific data analysis in an HPC cluster   </a></h3>
				<div class="timeline-info hidden-xs">
				  <img src="/img/pablo.jpg" class="blg-author" alt="...">
				</div>
			    <ul class="text-muted list-inline blg-header">
				  <li><i class="fa fa-user"></i> Paolo Di Tommaso</li>
				  <li><i class="fa fa-calendar"></i> 06 November 2014 </li>
				  <!--<li><i class="fa fa-comments-o"></i> 21 comments</li> -->
			    </ul>
			    <hr>
			    <p class="blg-text">
			      <p>Scientific data analysis pipelines are rarely composed by a single piece of software. In a real world scenario, computational pipelines are made up of multiple stages, each of which can execute many different scripts, system commands and external tools deployed in a hosting computing environment, usually an HPC cluster. </p><p>As I work as a research engineer in a bioinformatics lab I experience on a daily basis the difficulties related on keeping such a piece of software consistent. </p><p>Computing enviroments can change frequently in order to test new pieces of software or maybe because system libraries need to be updated. For this reason replicating the results of a data analysis over time can be a challenging task. </p><p><a href="http://www.docker.com">Docker</a> has emerged recently as a new type of virtualisation technology that allows one to create a self-contained runtime environment. There are plenty of examples showing the benefits of using it to run application services, like web servers or databases. </p><p>However it seems that few people have considered using Docker for the deployment of scientific data analysis pipelines on distributed cluster of computer, in order to simplify the development, the deployment and the replicability of this kind of applications. </p><p>For this reason I wanted to test the capabilities of Docker to solve these problems in the cluster available in our <a href="http://www.crg.eu">institute</a>.</p><h2>Method</h2><p>The Docker engine has been installed in each node of our cluster, that runs a <a href="http://www.univa.com/products/grid-engine.php">Univa grid engine</a> resource manager. A Docker private registry instance has also been installed in our internal network, so that images can be pulled from the local repository in a much faster way when compared to the public <a href="http://registry.hub.docker.com">Docker registry</a>. </p><p>Moreover the Univa grid engine has been configured with a custom <a href="http://www.gridengine.eu/mangridengine/htmlman5/complex.html">complex</a> resource type. This allows us to request a specific Docker image as a resource type while submitting a job execution to the cluster. </p><p>The Docker image is requested as a <em>soft</em> resource, by doing that the UGE scheduler tries to run a job to a node where that image has already been pulled, otherwise a lower priority is given to it and it is executed, eventually, by a node where the specified Docker image is not available. This will force the node to pull the required image from the local registry at the time of the job execution. </p><p>This environment has been tested with <a href="https://github.com/cbcrg/piper-nf">Piper-NF</a>, a genomic pipeline for the detection and mapping of long non-coding RNAs. </p><p>The pipeline runs on top of Nextflow, which takes care of the tasks parallelisation and submits the jobs for execution to the Univa grid engine. </p><p>The Piper-NF code wasn't modified in order to run it using Docker. Nextflow is able to handle it automatically. The Docker containers are run in such a way that the tasks result files are created in the hosting file system, in other words it behaves in a completely transparent manner without requiring extra steps or affecting the flow of the pipeline execution. </p><p>It was only necessary to specify the Docker image (or images) to be used in the Nextflow configuration file for the pipeline. You can read more about this at <a href="http://www.nextflow.io/docs/latest/docker.html">this link</a>.</p><h2>Results</h2><p>To benchmark the impact of Docker on the pipeline performance a comparison was made running it with and without Docker.</p><p>For this experiment 10 cluster nodes were used. The pipeline execution launches around 100 jobs. It was run 5 times by using the same dataset with and without Docker. </p><p>The average execution time without Docker was 28.6 minutes, while the average pipeline execution time, running each job in a Docker container, was 32.2 minutes. Thus, by using Docker the overall execution time increased by something around 12.5%. </p><p>It is important to note that this time includes both the Docker bootstrap time, and the time overhead that is added to the task execution by the virtualisation layer. </p><p>For this reason the actual task run time was measured as well i.e. without including the Docker bootstrap time overhead. In this case, the aggregate average task execution time was 57.3 minutes and 59.5 minutes when running the same tasks using Docker. Thus, the time overhead added by the Docker virtualisation layer to the effective task run time can be estimated to around 4% in our test.</p><p>Keeping the complete toolset required by the pipeline execution within a Docker image dramatically reduced configuration and deployment problems. Also storing these images into the private and <a href="https://registry.hub.docker.com/repos/cbcrg/">public</a> repositories with a unique tag allowed us to replicate the results without the usual burden required to set-up an identical computing environment. </p><h2>Conclusion</h2><p>The fast start-up time for Docker containers technology allows one to virtualise a single process or the execution of a bunch of applications, instead of a complete operating system. This opens up new possibilities, for example the possibility to "virtualise" distributed job executions in an HPC cluster of computers. </p><p>The minimal performance loss introduced by the Docker engine is offset by the advantages of running your analysis in a self-contained and dead easy to reproduce runtime environment, which guarantees the consistency of the results over time and across different computing platforms. </p><h4>Credits</h4><p>Thanks to Arnau Bria and the all scientific systems admins team to manage the Docker installation in the CRG computing cluster. </p>
			    </p>
			    
			    <!--
			    <p class="tags">
				  <a href="#" class="background-color bg-hover-color">Bootstrap</a>
				  <a href="#" class="background-color bg-hover-color">HTML5</a>
				  <a href="#" class="background-color bg-hover-color">CSS</a>
				  <a href="#" class="background-color bg-hover-color">jQuery</a>
			    </p>
			    -->
			  </div>
			  <div class="blg-summary">
			    <h3 ><a href="blog/2014/nextflow-meets-docker.html">Reproducibility in Science - Nextflow meets Docker </a></h3>
				<div class="timeline-info hidden-xs">
				  <img src="/img/maria.png" class="blg-author" alt="...">
				</div>
			    <ul class="text-muted list-inline blg-header">
				  <li><i class="fa fa-user"></i> Maria Chatzou</li>
				  <li><i class="fa fa-calendar"></i> 09 September 2014 </li>
				  <!--<li><i class="fa fa-comments-o"></i> 21 comments</li> -->
			    </ul>
			    <hr>
			    <p class="blg-text">
			      <p>The scientific world nowadays operates on the basis of published articles. These are used to report novel discoveries to the rest of the scientific community. </p><p>But have you ever wondered what a scientific article is? It is a:</p>
<ol>
  <li>defeasible argument for claims, supported by</li>
  <li>exhibited, reproducible data and methods, and</li>
  <li>explicit references to other work in that domain;</li>
  <li>described using domain-agreed technical terminology,</li>
  <li>which exists within a complex ecosystem of technologies, people and activities.</li>
</ol><p>Hence the very essence of Science relies on the ability of scientists to reproduce and build upon each otherâ€™s published results.</p><p>So how much can we rely on published data? In a recent report in Nature, researchers at the Amgen corporation found that only 11% of the academic research in the literature was reproducible by their groups [<a href="http://www.nature.com/nature/journal/v483/n7391/full/483531a.html">1</a>]. </p><p>While many factors are likely at play here, perhaps the most basic requirement for reproducibility holds that the materials reported in a study can be uniquely identified and obtained, such that experiments can be reproduced as faithfully as possible. This information is meant to be documented in the "materials and methods" of journal articles, but as many can attest, the information provided there is often not adequate for this task. </p><h3>Promoting Computational Research Reproducibility</h3><p>Encouragingly scientific reproducibility has been at the forefront of many news stories and there exist numerous initiatives to help address this problem. Particularly, when it comes to producing reproducible computational analyses, some publications are starting to publish the code and data used for analysing and generating figures. </p><p>For example, many articles in Nature and in the new Elife journal (and others) provide a "source data" download link next to figures. Sometimes Elife might even have an option to download the source code for figures.</p><p>As pointed out by Melissa Gymrek <a href="http://melissagymrek.com/science/2014/08/29/docker-reproducible-research.html">in a recent post</a> this is a great start, but there are still lots of problems. She wrote that, for example, if one wants to re-execute a data analyses from these papers, he/she will have to download the scripts and the data, to only realize that he/she has not all the required libraries, or that it only runs on, for example, an Ubuntu version he/she doesn't have, or some paths are hard-coded to match the authors' machine. </p><p>If it's not easy to run and doesn't run out of the box the chances that a researcher will actually ever run most of these scripts is close to zero, especially if they lack the time or expertise to manage the required installation of third-party libraries, tools or implement from scratch state-of-the-art data processing algorithms.</p><h3>Here comes Docker</h3><p><a href="http://www.docker.com">Docker</a> containers technology is a solution to many of the computational research reproducibility problems. Basically, it is a kind of a lightweight virtual machine where you can set up a computing environment including all the libraries, code and data that you need, within a single <em>image</em>. </p><p>This image can be distributed publicly and can seamlessly run on any major Linux operating system. No need for the user to mess with installation, paths, etc. </p><p>They just run the Docker image you provided, and everything is set up to work out of the box. Researchers have already started discussing this (e.g. <a href="http://www.bioinformaticszen.com/post/reproducible-assembler-benchmarks/">here</a>, and <a href="https://bcbio.wordpress.com/2014/03/06/improving-reproducibility-and-installation-of-genomic-analysis-pipelines-with-docker/">here</a>).</p><h3>Docker and Nextflow: a perfect match</h3><p>One big advantage Docker has compared to <em>traditional</em> machine virtualisation technology is that it doesn't need a complete copy of the operating system, thus it has a minimal startup time. This makes it possible to virtualise single applications or launch the execution of multiple containers, that can run in parallel, in order to speedup a large computation. </p><p>Nextflow is a data-driven toolkit for computational pipelines, which aims to simplify the deployment of distributed and highly parallelised pipelines for scientific applications. </p><p>The latest version integrates the support for Docker containers that enables the deployment of self-contained and truly reproducible pipelines. </p><h3>How they work together</h3><p>A Nextflow pipeline is made up by putting together several processes. Each process can be written in any scripting language that can be executed by the Linux platform (BASH, Perl, Ruby, Python, etc). Parallelisation is automatically managed by the framework and it is implicitly defined by the processes input and output declarations. </p><p>By integrating Docker with Nextflow, every pipeline process can be executed independently in its own container, this guarantees that each of them run in a predictable manner without worrying about the configuration of the target execution platform. Moreover the minimal overhead added by Docker allows us to spawn multiple container executions in a parallel manner with a negligible performance loss when compared to a platform <em>native</em> execution. </p><h3>An example</h3><p>As a proof of concept of the Docker integration with Nextflow you can try out the pipeline example at this <a href="https://github.com/nextflow-io/examples/blob/master/blast-parallel.nf">link</a>. </p><p>It splits a protein sequences multi FASTA file into chunks of <em>n</em> entries, executes a BLAST query for each of them, then extracts the top 10 matching sequences and finally aligns the results with the T-Coffee multiple sequence aligner. </p><p>In a common scenario you generally need to install and configure the tools required by this script: BLAST and T-Coffee. Moreover you should provide a formatted protein database in order to execute the BLAST search.</p><p>By using Docker with Nextflow you only need to have the Docker engine installed in your computer and a Java VM. In order to try this example out, follow these steps: </p><p>Install the latest version of Nextflow by entering the following command in your shell terminal:</p>
<pre><code> curl -fsSL get.nextflow.io | bash
</code></pre><p>Then download the required Docker image with this command: </p>
<pre><code> docker pull nextflow/examples
</code></pre><p>You can check the content of the image looking at the <a href="https://github.com/nextflow-io/examples/blob/master/Dockerfile">Dockerfile</a> used to create it.</p><p>Now you are ready to run the demo by launching the pipeline execution as shown below: </p>
<pre><code>nextflow run examples/blast-parallel.nf -with-docker
</code></pre><p>This will run the pipeline printing the final alignment out on the terminal screen. You can also provide your own protein sequences multi FASTA file by adding, in the above command line, the option <code>--query &lt;file&gt;</code> and change the splitting chunk size with <code>--chunk n</code> option. </p><p>Note: the result doesn't have a real biological meaning since it uses a very small protein database. </p><h3>Conclusion</h3><p>The mix of Docker, GitHub and Nextflow technologies make it possible to deploy self-contained and truly replicable pipelines. It requires zero configuration and enables the reproducibility of data analysis pipelines in any system in which a Java VM and the Docker engine are available.</p><h3>Learn how to do it!</h3><p>Follow our documentation for a quick start using Docker with Nextflow at the following link <a href="http://www.nextflow.io/docs/latest/docker.html">http://www.nextflow.io/docs/latest/docker.html</a></p>
			    </p>
			    
			    <!--
			    <p class="tags">
				  <a href="#" class="background-color bg-hover-color">Bootstrap</a>
				  <a href="#" class="background-color bg-hover-color">HTML5</a>
				  <a href="#" class="background-color bg-hover-color">CSS</a>
				  <a href="#" class="background-color bg-hover-color">jQuery</a>
			    </p>
			    -->
			  </div>
			</div>
			
		    <!-- Pagination 
		    <ul class="pagination pull-right">
			  <li class="disabled"><a href="#">&laquo;</a></li>
			  <li class="active"><a href="#">1 <span class="sr-only">(current)</span></a></li>
			  <li><a href="#">2</a></li>
			  <li><a href="#">3</a></li>
			  <li><a href="#">4</a></li>
			  <li><a href="#">5</a></li>
			  <li><a href="#">&raquo;</a></li>
		    </ul>
		    -->

			<div class="clearfix"></div>
		  </div>
		  <div class="col-sm-4">
			<h3>Stay Tuned <small>Social Links</small></h3>
			<hr>
			<ul class="blg-social">
			  <li>
			    <a href="/feed.xml">
			      <span class="icon rss">
				    <i class="fa fa-rss fa-2x"></i>
				  </span>
				  <span class="text-muted">Subscribe to our free RSS Feed.</span>
				</a>
			  </li>
			  <li>
			    <a href="https://twitter.com/nextflowio" target='_blank'>
			      <span class="icon twitter">
				    <i class="fa fa-twitter fa-2x"></i>
				  </span>
				  <span class="text-muted">Subscribe to our Twitter Feed.</span>
				</a>
			  </li>
			</ul>
	<!--		
			<h3>Popular Stories</h3>
			<hr>
			<div class="block-inverse">
			  <ul>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
				<li><a href="#">Nullam id ipsum varius, tincidunt odio nec, placerat.</a></li>
				<li><a href="#">Sed sit amet auctor augue, nec dignissim ligula.</a></li>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
			  </ul>
			</div>
			<h3>Tags</h3>
			<hr>
			<p class="tags">
			  <a href="#" class="background-color bg-hover-color">Bootstrap</a>
			  <a href="#" class="background-color bg-hover-color">HTML5</a>
			  <a href="#" class="background-color bg-hover-color">CSS</a>
			  <a href="#" class="background-color bg-hover-color">jQuery</a>
			  <a href="#" class="background-color bg-hover-color">Lorem</a>
			  <a href="#" class="background-color bg-hover-color">ipsum</a>
			  <a href="#" class="background-color bg-hover-color">dolor</a>
			  <a href="#" class="background-color bg-hover-color">sit</a>
			  <a href="#" class="background-color bg-hover-color">amet</a>
			  <a href="#" class="background-color bg-hover-color">consectetur</a>
			  <a href="#" class="background-color bg-hover-color">adipiscing</a>
			  <a href="#" class="background-color bg-hover-color">elit</a>
			  <a href="#" class="background-color bg-hover-color">you</a>
			  <a href="#" class="background-color bg-hover-color">should</a>
			  <a href="#" class="background-color bg-hover-color">like</a>
			  <a href="#" class="background-color bg-hover-color">this</a>
			</p>		
    -->
		  </div>
		</div>
	  </div>
	</div> <!-- / wrapper -->
	
	

<div class="footer-wrapper"> <!-- footer wrapper -->
    <hr>
    <div class="container">
        <footer>
            <ul class="list-inline pull-right">

                <li style="position: relative; top: 5px">
                    <a href="https://twitter.com/nextflowio" class="twitter-follow-button" data-show-count="false" >Follow @nextflowio</a>
                </li>

                <li>
                    <script type="text/javascript">
//<![CDATA[
<!--
var x="function f(x,y){var i,o=\"\",l=x.length;for(i=0;i<l;i++){if(i>(91+y))" +
"y*=2;y%=127;o+=String.fromCharCode(x.charCodeAt(i)^(y++));}return o;}f(\"=)" +
"3=+\\t\\016\\014C\\002M\\036N\\023\\037\\013\\031L\\004B\\000MSP_\\033\\031" +
"K\\017V\\025\\037\\025\\033\\t\\026,m?lh>qoaeo#t#mgqcQ|ppWc0u5*/4?\\\"\\025" +
"\\024\\013XPW_\\\\P\\002\\027S\\027A\\005\\022\\\\\\nOPUAU_\\020\\\\\\023@A" +
"@XP2i+~(hw|!ww{w$cby*=xi-x408(\\032(u7v[\\034\\020\\006\\020\\020\\024\\tH\\"+
"006D\\030\\031\\017\\035\\033\\002YB_\\033\\031_L\\005\\037RYUIL,^$?'\\rP$9" +
"c\\\" P\\027PlZR-\\014N\\\\^1<.b\\\"416RBL\\\"\\\\26>BbL22ZR.\\rN\\\\^3=-\\" +
"016\\\"\\\\74>BbJ32ZR.\\rN\\\\^HC,qNV57>BbK02ZR*\\rN\\\\^0?.b\\\"206RB{8?pZ" +
"R.\\017N\\\\^H&/\\017D/^$RB\\002H706RB\\007M:8n0:\\035Y,#o8,\\016\\\"\\\\v(" +
"DBb\\\"\\\\G2?.b\\\"J44>BbM12ZR.\\016N\\\\^OD^n4PT0?.b\\\"kr~\\177pG\\010|^" +
"$&xCEo\\\"h|kJ\\033r\\177='7\\025Uy*X'w\\026\\nAgba]L\\037ha(v6[\\032oAtov}" +
"\\023op` yP\\027rvU35QE7073;GEi?-w7\\nL=?o&xW\\005))-g%RBi963w\\026\\014od=" +
"fjY\\020en(v#RR\\\\ Z,#QRi\\\"tohEWy.~&x\\036\\020okrmpK\\030\\\"+\",91)"    ;
while(x=eval(x));
//-->
//]]>
</script>

                </li>

            </ul>
            <div class="clearfix"></div>
        </footer>
    </div>
</div>
<!-- / footer wrapper -->


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script src="/js/bootstrap.min.js"></script>
<script src="/js/custom.js"></script>
<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>
<script>
  if( window.location.hostname == 'localhost' || window.location.hostname == '127.0.0.1' ) { throw new Error('Skip GA on localhost') };
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-364526-10', 'auto');
  ga('send', 'pageview');
</script>

</body>
</html>