---
title: "How I became a Nextflow Ambassador!"
date: 2024-07-24
type: post
description: This blog describes my journey from bash to nextflow community.
image: /img/blog-2024-07-24--share.png
tags: nextflow,ambassador_post
status: published
author: Firas Zemzem
icon: FirasZemzemblogIcon.jpeg
community_post: true
ambassador_post: true
---

As a PhD student in bioinformatics, I aimed to build robust pipelines to analyze diverse datasets throughout my research. Initially, mastering Bash scripting was a time-consuming challenge, but this journey ultimately led me to become a Nextflow Ambassador, engaging actively with the expert Nextflow community.

<!-- end-archive-description -->

My name is [Firas Zemzem](https://www.linkedin.com/in/firaszemzem/), a PhD student based in [Tunisia](https://www.google.com/search?q=things+to+do+in+tunisia&sca_esv=3b07b09e3325eaa7&sca_upv=1&udm=15&biw=1850&bih=932&ei=AS2eZuqnFpG-i-gPwciJyAk&ved=0ahUKEwiqrOiRsbqHAxUR3wIHHUFkApkQ4dUDCBA&uact=5&oq=things+to+do+in+tunisia&gs_lp=Egxnd3Mtd2l6LXNlcnAiF3RoaW5ncyB0byBkbyBpbiB0dW5pc2lhMgUQABiABDIGEAAYFhgeMgYQABgWGB4yBhAAGBYYHjIGEAAYFhgeMgYQABgWGB4yBhAAGBYYHjIGEAAYFhgeMgYQABgWGB4yCBAAGBYYHhgPSOIGULYDWNwEcAF4AZABAJgBfaAB9gGqAQMwLjK4AQPIAQD4AQGYAgOgAoYCwgIKEAAYsAMY1gQYR5gDAIgGAZAGCJIHAzEuMqAH_Aw&sclient=gws-wiz-serp) working with the Laboratory of Cytogenetics, Molecular Genetics, and Biology of Reproduction at CHU Farhat Hached Sousse. I was specialized in human genetics, focusing on studying genomics behind neurodevelopmental disorders. Hence Developing methods for detecting SNPs and variants related to my work was crucial step for advancing medical research and improving patient outcomes. On the other hand, pipelines integration and bioinformatics tools were essential in this process, enabling efficient data analysis, accurate variant detection, and streamlined workflows that enhance the reliability and reproducibility of our findings.

## The initial nightmare of Bash

During my master's degree, I was a steadfast user of Bash scripting. Bash had been my go-to tool for automating tasks and managing workflows in my bioinformatics projects, such as variant calling. Its simplicity and versatility made it an indispensable part of my toolkit. I was writing Bash scripts for various next-generation sequencing (NGS) high-throughput analyses, including data preprocessing, quality control, alignment, and variant calling. However, as my projects grew more complex, I began to encounter the limitations of Bash. Managing dependencies, handling parallel executions, and ensuring reproducibility became increasingly challenging. Handling the vast amount of data generated by NGS and other high-throughput technologies was cumbersome. Using Bash became a nightmare for debugging and maintaining. I spent countless hours trying to make it work, only to be met with more errors and inefficiencies. It was nearly impossible to scale for larger datasets and more complex analyses. Additionally, managing different environments and versions of tools was beyond Bash's capabilities. I needed a solution that could handle these challenges more gracefully.

## Game-Changing Call

One evening, I received a call from my friend, Mr. HERO, a bioinformatician. As we discussed our latest projects, I vented my frustrations with Bash. Mr. HERO, as I called him, the problem-solver, mentioned a tool called Nextflow. He described how it had revolutionized his workflow, making complex pipeline management a breeze. Intrigued, I decided to look into it.

## Diving Into the process

Reading the [documentation](https://www.nextflow.io/docs/latest/index.html) and watching [tutorials](https://training.nextflow.io/) were my first steps. Nextflow's approach to workflow management was a revelation. Unlike Bash, Nextflow was designed to address the complexities of modern computational questions. It provided a transparent, declarative syntax for defining tasks and their dependencies and supported parallel execution out of the box. The first thing I did when I decided to convert one of my existing Bash scripts into a Nextflow pipeline was to start experimenting with simple code. Doing this was no small feat. I had to rethink my approach to workflow design and embrace a new way of defining tasks and dependencies. My learning curve was not too steep, so understanding how to translate my Bash logic into Nextflow's domain-specific language (DSL) was not that hard.

## Eureka Moment: First run

The first time I ran my Nextflow pipeline, I was amazed by how smoothly and efficiently it handled tasks that previously took hours to debug and execute in Bash. Nextflow managed task dependencies, parallel execution, and error handling with ease, resulting in a faster, more reliable, and maintainable pipeline. The ability to run pipelines on different computing environments, from local machines to high-performance clusters and cloud platforms, was a game-changer. Several Nextflow features were particularly valuable: Containerization Support using Docker and Singularity ensured consistency across environments; Error Handling with automatic retry mechanisms and detailed error reporting saved countless debugging hours; Portability and scalability allowed seamless execution on various platforms; Modularity facilitated the reuse and combination of processes across different pipelines, enhancing efficiency and organization; and Reproducibility features, including versioning and traceability, ensured that workflows could be reliably reproduced and shared across different research projects and teams.

<div style="margin-top: 2rem; margin-bottom: 2rem;">
    <img src="/img/ZemFiras-nextflowtestpipeline-Blog.png" alt="meme on bright landscape" />
</div>

## New Horizons: Becoming a Nextflow Ambassador

Switching from Bash scripting to Nextflow was more than just adopting a new tool. It was about embracing a new mindset. Nextflowâ€™s emphasis on scalability, reproducibility, and ease of use transformed how I approached bioinformatics. The initial effort to learn Nextflow paid off in spades, leading to more robust, maintainable, and scalable workflows. My enthusiasm and advocacy for Nextflow didn't go unnoticed. Recently, I became a Nextflow Ambassador. This role allows me to further contribute to the community, promote best practices, and support new users as they embark on their own Nextflow journeys.

## Future Projects and Community Engagement

Currently I am working on developing a Nextflow pipeline with my team that will help in analyzing variants, providing valuable insights for medical and clinical applications. This pipeline aims to improve the accuracy and efficiency of variant detection, ultimately supporting better diagnostic for patients with various genetic conditions. As part of my ongoing efforts within the Nextflow community, I am planning a series of projects aimed at developing and sharing advanced Nextflow pipelines tailored to specific genetic rare disorder analyses. These initiative will include detailed tutorials, case studies, and collaborative efforts with other researchers to enhance the accessibility and utility of Nextflow for various bioinformatics applications. Additionally, I plan to host workshops and seminars to spread knowledge and best practices among my colleagues and other researchers. This will help foster a collaborative environment where we can all benefit from the power and flexibility of Nextflow.

## Invitation for researchers over the world

As a Nextflow Ambassador, I invite you to become part of a dynamic group of experts and enthusiasts dedicated to advancing workflow automation. Whether you're just starting or looking to deepen your knowledge, our community offers invaluable resources, support, and networking opportunities. You can chat with us on the [Nextflow Slack Workspace](https://www.nextflow.io/slack-invite.html) and ask your questions at the [Seqera Community Forum](https://community.seqera.io).
